### PR TITLE
NAS-113060 / 22.02-RC.2 / Allow mocking methods in integration tests

### PR DESCRIPTION
**Pull request by themylogin**

### CODE REVIEW

**Comment by themylogin** [comment]
time 12:00

**Comment by themylogin** [comment]
time 2:30

File: tests/api2/test_377_reporting.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import time
+
+from middlewared.utils.test import *
```
[the remark this answers is not in the document]
Comment (ericbsd) [reply tests/api2/test_377_reporting.py:3]: tests/api2/ is for REST and WebSocket API, I am not a fan of installing middleware on my local system to run the API test.

IMHO tests/api2/test_377_reporting.py should be implemented somewhere else.

[the remark this answers is not in the document]
Comment (ericbsd) [reply tests/api2/test_377_reporting.py:3]: There is already code added for WebSocket API look at https://github.com/truenas/middleware/blob/1c23415473361c2642c2db9afbcd7abc6f130df4/tests/functions.py#L265-L281

File: tests/runtest.py [hunk 2]
```diff
@@ -100,6 +101,8 @@
         dev_test = True
     elif output == '--debug-mode':
         debug_mode = True
+    elif output == '-v':
```
[the remark this answers is not in the document]
Comment (ericbsd) [reply tests/runtest.py:104]: what are the arguments for -v?

File: tests/api2/test_377_reporting.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (themylogin) [reply tests/api2/test_377_reporting.py:3]: @ericbsd this is too low-level, I need middleware client library

File: tests/runtest.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (themylogin) [reply tests/runtest.py:104]: No arguments, just `-v`, the same way as `pytest` itself uses it

File: tests/api2/test_377_reporting.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ericbsd) [reply tests/api2/test_377_reporting.py:3]: Could we find a way to run that test directly in TrueNAS?

[the remark this answers is not in the document]
Comment (ericbsd) [reply tests/api2/test_377_reporting.py]: @themylogin I think the code below would work for systems that do not have the middleware client library.
```
try:
    from middlewared.utils.test import *
except ImportError:
    Reason = 'middleware client library not installed'
    pytestmark = pytest.mark.skip(reason=Reason)
```
Basically, if the middlewared.utils.test is not present pytestmark will skip all tests in the file.

**Review by ericbsd — APPROVED** [review]
LGTM

**Pull request merged**